### ISSUE TITLE
Type inference failed for class attributes in lambda

### ISSUE THREAD

**Issue by ColinTree** [palantir/tslint#4787]
### Bug Report

- __TSLint version__: 5.18.0
- __TypeScript version__: 3.5.2
- __Running TSLint via__: VSCode

#### Reproduction using [TSLint Playground](https://palantir.github.io/tslint-playground)

Cannot be reproduced by tslint-playground (it magically works)

I had tried to run `npm update`  to ensure tslint & typescript is the latest version, but it still behaves differently with the tslint-playground

[url to playground](https://palantir.github.io/tslint-playground/?saved=N4Igxg9gJgpiBcID0SAEAXAzgGwJYDt14pdMBDAI2xlTICc6IB3AWgAd6Z9NV8IXI3CNQA6+MegCebGgHkKAKxhh0sugDkArtmyoAvKmCoA2gGsYk+Kkzo6BAOYBdKzbv57qAL6oAPr23YANxiYmDYZJg8AGIQEIYh+KiobHYAbmToNOnYmjBW8koqalo6+oaeweKJyWkZNABC9KgAFACU8dVJuABmLegAFqQAdNm5+noG+AHtwGJJ86goqLgAtmwQkbhUMHPz3jDYmDSznfNLWHiEqGBkibCZdCsEWWQ5NKQY0scm5pbWtg5nP83B5vM0+PgWFMdK1dgsCsp0ENfphmgNhqMYLDTgsFkNuhA6ABRMhgfrNX76AB8GEGmBGr1yZgsjlagUWaCJDEJqCYuAGqBEIHR9MxQtQKzIklQFBo0OwcLOaAGNAg+Gw0swwk06FwavgiqSCJUyIsqJFDLe2NxNtQ+MJJLJFIs1L6dMtYwihh+FhcAPcQNcDi8rWZklZlQWnl20fwsbESwJOmYwaYhNMPG6zwSghsqEx+UUiOKATKwAqCR6LUx40m0w6SSWq3Wm22Yn2h2Ou1zwhgQ2wEHszUVxqRKOamOttqS9uJpPJlL0NMxYdZbI5vDiXMYdHbYhAABpwGqs-YECATgshTAAB6ZfBQTBCqzGIUXAhEOjKCArFZcWBQEKjgHoqQp0NoMBPgg5R7vgICeEAA)

#### TypeScript code being linted

```ts
// tslint:disable no-console

type ObjectOrNull = { [key: string]: string } | null;

class Foo {

  private value: ObjectOrNull = {};

  private Bar () {
    if (this.value === null) {
      // impossible
    } else {
      // tslint can determine value is type { [key: string]: string } (non-null)
      Object.keys(this.value)
            .forEach(key => this.value[key]); // Error with "this.value" may be null
      // the only solution:
      Object.keys(this.value)
            .forEach(key => (this.value as { [key: string]: string })[key]);
    }
  }
}

// following works fine

const value: ObjectOrNull = {};

if (value === null) {
  // impossible
} else {
  console.log(
    Object.keys(value)
          .forEach(key => value[key])); // no Error
}
```

with `tslint.json` configuration:

```json
{
  "defaultSeverity": "warning",
  "extends": [
    "tslint:recommended"
  ],
  "linterOptions": {
    "exclude": [
      "node_modules/**"
    ]
  },
  "rules": {
    "forin": false,
    "space-before-function-paren": true,
    "arrow-parens": false,
    "quotemark": [true, "single"],
    "indent": [true, "spaces", 2],
    "interface-name": false,
    "ordered-imports": false,
    "object-literal-sort-keys": false,
    "no-consecutive-blank-lines": false
  }
}
```

#### Actual behavior & Expected behavior

See comments in code

**Comment by adidahiya** [comment]
Which lint rule is actually producing an error? This seems like it may be a TS compilation issue, not TSLint

**Comment by JoshuaKGoldberg** [comment]
@ColinTree it looks like your issue is with **TypeScript**, not TSLint. TypeScript is the programming language and compiler that's giving you a type error in your code. TSLint is a set of analysis that uses TypeScript.

Specifically, your issue is that TypeScript doesn't preserve type narrowing within functions. Simplifying your code:

```ts
let something: string | null = "hi";

if (Math.random() > 0.5) something = null;

if (something) {
    /* Point A */
    setTimeout(() => { /* Point B */ console.log(something.length); }, 0);
} else {
    console.log("nothing going");
}
```

At `Point A`, TypeScript has _narrowed_ that `something` is definitely type `string`. It knows this because the `if` statement's body will only be hit if so. However, TypeScript intentionally forgets the narrowing in `Point B` because it's in a new function scope. 

If you find this or other TypeScript issues disagreeable, you'll want to file them on https://github.com/Microsoft/TypeScript/issues. Enjoy!
